src/library/scala/math/BigInt.scala

```diff
@@ -132,11 +212,13 @@ final class BigInt(val bigInteger: BigInteger)
     case that: Float      => isValidFloat && toFloat == that
     case x                => isValidLong && unifiedPrimitiveEquals(x)
   }
-  override def isValidByte: Boolean = this >= Byte.MinValue && this <= Byte.MaxValue
-  override def isValidShort: Boolean = this >= Short.MinValue && this <= Short.MaxValue
-  override def isValidChar: Boolean = this >= Char.MinValue && this <= Char.MaxValue
-  override def isValidInt: Boolean = this >= Int.MinValue && this <= Int.MaxValue
-           def isValidLong: Boolean = this >= Long.MinValue && this <= Long.MaxValue
+
+  override def isValidByte: Boolean = longEncoding && _long >= Byte.MinValue && _long <= Byte.MaxValue
+  override def isValidShort: Boolean = longEncoding && _long >= Short.MinValue && _long <= Short.MaxValue
+  override def isValidChar: Boolean = longEncoding && _long >= Char.MinValue && _long <= Char.MaxValue
+  override def isValidInt: Boolean = longEncoding && _long >= Int.MinValue && _long <= Int.MaxValue
+           def isValidLong: Boolean = longEncoding || _bigInteger.bitLength() <= 63 // rhs of || tests == Long.MinValue
```

Comment:
why not more direct: `longEncoding || (_bigInteger == BigInt.longMinValueAsBigInteger)`

Reply:
I'm very skeptical of being defensive against invariant violations. It's very hard to reason about what else may go wrong in such cases. I think we should test well enough, and encapsulate strongly enough that we are very very sure the invariants are not violated.

Reply:
Changed in the direction of @johnynek suggestion